### DISCUSSION TITLE
Any way to access local symbol table of calling function?

### DISCUSSION THREAD

**Discussion by farhansolodev** [GoogleFeud/ts-macros#82] [Q&A] [answered]
In python, you can do something like:
```python
inspect.currentframe().f_back.f_locals
```
to get a dict of local variables and their values from the calling scope.

Does ts-macros enable similiar functionality? Something like `$$typeMetadata` but for a frame instead of a type.

**Comment by farhansolodev** [comment]
Found more suitable methods to achieve this

**Comment by volenslavchev** [comment]
Hi! You can do this with the `$$raw` built-in macro:

```ts
    const locals = $$raw!((ctx) => {
        const variableSymbols = [...ctx.thisMacro.macro.node.locals.values()];
        return ctx.factory.createArrayLiteralExpression(variableSymbols.map(l => ctx.factory.createIdentifier(l.name)));
    });
```

[playground](https://googlefeud.github.io/ts-macros/?code=FAehAIEkDsBcFMBOBDAxrAlgN3uADgDbICeA5ogPYCu0AJgISgQDqiGC4x1i4qFtuABZJcyOuADO8XLGHhYKaBLwYC8WuETwJVArAB0TcAEECBecTzbwYjQDMa6DBSXg7lALby5BDACMURGIbLRsCLWRaYIwPPApEBAZgYAdoJxdwABIECVgACgBKcABvYHBy3hdcm3AAXnAARgBuMoq+JVhwPzrwACYWisqO8AIKVGQCCR7MzJQAd3o8vPQADyLagD4S1sHy9uqsZDZkPzUAZWIPPwpJnoBtfUfV-VkMCQBZNEp9Dy+KfWg-Hg+lG40m+kOBCo2kKAF0BrtylpYFRENBeLAVvo7GhYPFiPpUBEEMZEChiAAZdhICYAURWeC0EgkzmgeUOx1O8AuVxuEh+yDweXMm22iMRyNR6OeOPQ+MJxPgkAEcAwdgwSGFAOQHngBQRiIAvgV9TtwMaWobksBstpYIt9UA)

**Reply by farhansolodev** [reply]
I see you define the `Macro.node` as `ts.FunctionDeclaration`. I'm not familiar with this type; Does `node` have other things I can access? Particularly, I want to know the names of the variables passed into the macro function (so that users can just pass variables into the macro instead of a string that represents the name of the variable they want to pass in). Is this possible?

**Reply by volenslavchev** [reply, marked as answer]
I'm not sure I understand, this is the default behaviour. Say you pass the variable `a` to a macro whose first parameter is called `b`. When the macro is expanding, all occurrences of `b` inside the macro body will be replaced with `a`. [Here's an example](https://googlefeud.github.io/ts-macros/?code=FAMwrgdgxgLglgewgAgCQFsCeBZAhlAJwQAoAjASmQG9hk7kCBTGMAlVVGRgDxgEIylANTIAjAG5gAX2DAANs2S5kAXjGTgUJAGcYDRtrBy9ajDnxEBucuKA)

**Reply by farhansolodev** [reply]
Oh this works perfectly 😅. Now I can do this:

```ts
class Prop5 {
    prop1
    prop2
    prop3
    prop420
    prop1337
    constructor() {
        this.prop1 = true;
        this.prop2 = "its a string";
        this.prop3 = 0xBabeCafe;
        this.prop32 = 32;
        this.prop1337 = new Foo2(3, 2);
    }

    from(prop) {
        $copy!<this>(this, prop)
    }
}
```
with
```ts
function $copy<F extends Record<any, any>>(to: Record<any, any>, from: F) {
    +[[$$typeMetadata!<F>(true).properties], (prop: TypeMetadataProperty) => {
        to[prop.name] = from[prop.name]
    }];
}
```

to get
```ts
    from(prop) {
        this["prop1"] = prop["prop1"];
        this["prop2"] = prop["prop2"];
        this["prop3"] = prop["prop3"];
        this["prop32"] = prop["prop32"];
        this["prop1337"] = prop["prop1337"];
    }
```

...although I do wish I could just do `$copy!(this, prop)` without the generic argument and $$typeMetadata would just infer that it's Prop5

**Reply by volenslavchev** [reply]
You can infer it by using 2 type parameters:
```ts
function $copy<T, F extends Record<any, any>>(to: T, from: F) {
    +[[$$typeMetadata!<T>(true).properties], (prop: TypeMetadataProperty) => {
        to[prop.name] = from[prop.name]
    }];
}
```

**Reply by farhansolodev** [reply]
While this isn't what I want because I want to enumerate properties from `from` not `to`, I realised why I need the generic argument in my case: its because the `from(prop)` definition in my example doesnt specify any type for `prop`, so ofcourse $$typeMetadata will be unable to enumerate `from`'s props, it doesnt have any type lol. 

Just changing the definition to `from(prop: Prop5)` allows me to omit the generic argument to my original $copy function. My bad for the confusion, and thanks for the help 😊.
